ONSdigital/census-rh-service
# Cr 118 receive case events from rm

# Motivation and Context

Receives CaseCreated and CaseUpdated type messages and deserialises them into relevant Java objects (CaseEvent objects) and then stores them, as JSON, in a bucket, named case_bucket, within the Google Cloud Platform. Also receives UACUpdated type messages and deserialises them into UACEvent objects and then stores them, as JSON, in a bucket (named uac_bucket) within the Google Cloud Platform.

# What has changed

Various Java classes have been added to the package named uk.gov.ons.ctp.integration.rhsvc.message.impl

JUnit tests have also been added or amended as appropriate (GCSDataStoreUnitTest, CaseEventReceiverImplTest, and UacEventReceiverImplTest)




# How to test?

Use the readme file to firstly set up Google credentials locally. Then run Rabbitmq locally using the command 'docker-compose up -d'. Use the following JSON format to put a CaseUpdated type message on the case-outbound-exchange (and use event.case.lifecycle as the routing key when you publish the message). You can then run the code and should see the message appear in Google Cloud Platform within a bucket named case_bucket NB. To see the buckets you need to select Storage --> Browser (within GCP). The code will create the bucket if it does not already exist (although there may be a problem if a bucket with the same name already exists in another project within GCP).

{
  "event" : {
    "type" : "CASE_UPDATED",
    "source" : "CASE_SERVICE",
    "channel" : "RM",
    "dateTime" : "2011-08-12T20:17:46.384Z",
    "transactionId" : "c45de4dc-3c3b-11e9-b210-d663bd873d93"
  },
  "payload" : {
    "collectionCase" : {
        "id":"bbd55984-0dbf-4499-bfa7-0aa4228700e9",
        "caseRef":"10000000010",
        "survey":"CENSUS",
        "collectionExerciseId":"n66de4dc-3c3b-11e9-b210-d663bd873d93",
        "address": {
		"addressLine1":"1 main street",
  		"addressLine2":"upper upperingham",
  		"addressLine3":"",
  		"townName":"upton",
  		"postcode":"UP103UP",
  		"region":"E",    
  		"latitude":"50.863849",
  		"longitude":"-1.229710",
  		"uprn":"XXXXXXXXXXXXX",
  		"arid":"XXXXX",
  		"addressType":"CE",
  		"estabType":"XXX"
	},
	"contact": {
	"title":"Ms",
  	"forename":"jo",
  	"surname":"smith",
  	"email":"me@example.com",
  	"telNo":"+447890000000"
	},
        "state":"ACTIONABLE",
        "actionableFrom":"2011-08-12T20:17:46.384Z"
}
  }
}

Then use the following JSON format to put a UACUpdated type message on the case-outbound-exchange (and use event.uac.updates as the routing key when you publish the message). You can then run the code and should see the message appear in Google Cloud Platform within a bucket named uac_bucket. 

{
  "event" : {
    "type" : "UAC_UPDATED",
    "source" : "CASE_SERVICE",
    "channel" : "RM",
    "dateTime" : "2011-08-12T20:17:46.384Z",
    "transactionId" : "c45de4dc-3c3b-11e9-b210-d663bd873d93"
  },
  "payload" : {
    "uac" : {
        "uacHash":  "72C84BA99D77EE766E9468A0DE36433A44888E5DEC4AFB84F8019777800B7364",        
        "active" : true,
        "questionnaireId":"1110000009",
        "caseType":"H",
        "region":"E",
        "caseId":"bbd55984-0dbf-4499-bfa7-0aa4228700e9",
        "collectionExerciseId":"n66de4dc-3c3b-11e9-b210-d663bd873d93"
}
  }
}
